pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,345 @@
+	podRequest, podLimits := getPodsTotalRequestsAndLimits(pods)
+	daemonRequest, daemonLimits := getPodsTotalRequestsAndLimits(daemonSetPods)
+	systemOverhead := getSystemOverhead(node)
+	allocatable := node.Status.Capacity
+	if len(node.Status.Allocatable) > 0 {
+		allocatable = node.Status.Allocatable
+	}
+	// Populate  metrics
+	for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+		overheadGaugeVec:       systemOverhead,
+		podRequestsGaugeVec:    podRequest,
+		podLimitsGaugeVec:      podLimits,
+		daemonRequestsGaugeVec: daemonRequest,
+		daemonLimitsGaugeVec:   daemonLimits,
+		allocatableGaugeVec:    allocatable,
+	} {
+		if err := c.insertGaugeValues(resourceList, node, gaugeVec); err != nil {
+			logging.FromContext(ctx).Errorf("Failed to generate gauge: %w", err)
+		}
+	}
+	return nil
+}
+
+func getSystemOverhead(node *v1.Node) v1.ResourceList {
+	systemOverheads := v1.ResourceList{}
+	if len(node.Status.Allocatable) > 0 {
+		// calculating system daemons overhead
+		for resourceName, quantity := range node.Status.Allocatable {
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
+			systemOverheads[resourceName] = overhead
+		}
+	}
+	return systemOverheads
+}
+
+func (c *Controller) insertGaugeValues(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+
+	for resourceName, quantity := range resourceList {
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		labels := c.generateLabels(node, resourceTypeName)
+		// Register the set of labels that are generated for node
+		nodeNamespacedName := types.NamespacedName{Name: node.Name}
+		c.LabelSliceMap[nodeNamespacedName] = append(c.LabelSliceMap[nodeNamespacedName], labels)
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generate new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()) / float64(1000))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+// GetPodsTotalRequestsAndLimits calculates the total resource requests and limits for the pods.
+// If pod overhead is non-nil, the pod overhead is added to the
+// total container resource requests and to the total container limits which have a non-zero quantity.
+func getPodsTotalRequestsAndLimits(pods []*v1.Pod) (reqs v1.ResourceList, limits v1.ResourceList) {
+	reqs, limits = v1.ResourceList{}, v1.ResourceList{}
+	for _, pod := range pods {
+		// Excluding pods that are completed or failed
+		if podutil.IsTerminal(pod) {
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			// Calculate Resource Requests
+			addResourceQuantity(container.Resources.Requests, reqs)
+			// Calculate Resource Limits
+			addResourceQuantity(container.Resources.Limits, limits)
+		}
+		// Add overhead for running a pod to the sum of requests and to non-zero limits:
+		if pod.Spec.Overhead != nil {
```

Comment:
IIRC we were going to skip Overhead and just use the existing `resources.RequestsForPods(pods), resources.LimitsForPods(pods)` helpers.

Reply:
done